### PR TITLE
faster_and_more_structural_diff

### PR DESCRIPTION
**Pull request by danieldroit**
1. use of clj-fast for faster data-structure ops.
2. added option for short-circuiting computation if the edit distance is above the given value.
3. added option for providing a custom comparator instead of always using =
4. Usage of transients for further speed improvements.
5. Replacing optimized java source code with equivalent fast idiomatic Clojure.

### CODE REVIEW

File: src/clj/clj_diff/core.clj [hunk 1]
```diff
@@ -17,27 +18,34 @@
...
+                              (if (= index -1)
+                                (assoc a 0 (fv/vector {::insert items} (get a 0)))
+                                (assoc a index (fv/vector (get a index) {::insert items})))))
+                    s
+                    additions)]
+    (mapcat
+      #(if (vector? %)
+         %
+         [%])
+      s)
```
Comment (edwardwas) [inline src/clj/clj_diff/core.clj]: This might be a bit "bike shed"y, but I'd find this easier to read without binding temporary variables. If it was me I'd reach for `as->`, but I do like my arrows. Feel free to ignore if you'd rather not

File: src/clj/clj_diff/core.clj [hunk 2]
```diff
@@ -17,27 +18,34 @@
...
+
+    #_(apply concat s)))
```
Comment (edwardwas) [inline src/clj/clj_diff/core.clj]: Leftover dev work?

File: src/clj/clj_diff/core.clj [hunk 3]
```diff
@@ -53,25 +61,39 @@
...
   [s edit-script]
-  (apply str (patch* s edit-script)))
+  (apply str
+    (mapcat
+      (fn [item]
+        (cond
+          (and (map? item) (contains? item :clj-diff.core/delete)) nil
+          (and (map? item) (contains? item :clj-diff.core/insert)) (:clj-diff.core/insert item)
+          :else (list item)))
+      (patch* s edit-script))))
```
Comment (edwardwas) [inline src/clj/clj_diff/core.clj:80]: Is this the same as the default case? If so it might be worth adding a comment here pointing that out. Just for ease of reading in the future.

File: src/clj/clj_diff/miller.clj [hunk 4]
```diff
@@ -41,31 +49,41 @@
...
+
+      (cond
+        (and limit-p
+             (> p limit-p))
+        #_=> [-1 delta fp]
+
+        (= (-> (fast/val-at fp (dec p) {})
+               (fast/val-at delta))
+           n)
+        #_=> [(dec p) delta fp]
```
Comment (edwardwas) [inline src/clj/clj_diff/miller.clj:82]: More commented out code

**Review by edwardwas — APPROVED** [review]
On the whole looks good. I've put some comments in about things that made it slightly hard to follow what the code was doing, but they're all optional - feel free to skip them if you don't think they're needed or worth the time.

My only other comment is that in `edit-distance` the form of `opts` isn't documented (or at least I couldn't see where). I'm not sure where the best place for it would be, but I think it'd be good to have a docstring or comment explaining what fields this can have

File: src/clj/clj_diff/core.clj (hunk 1 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj]: I appreciate the comment. I understand this comment regards the project as a whole and is not specific to this PR tho.

Comment (edwardwas) [reply src/clj/clj_diff/core.clj]: Yep, it's a bit "procedural" in places. If you wanted to change it now, one could. But don't feel the need to if you don't think it's worth it

File: src/clj/clj_diff/core.clj (hunk 2 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj]: yes, thanks

File: src/clj/clj_diff/miller.clj (hunk 4 above)
Comment (edwardwas) [reply src/clj/clj_diff/miller.clj:82]: Apparently I can't read properly and these are fine

File: src/clj/clj_diff/core.clj (hunk 3 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj:80]: the cases differ slightly

File: src/clj/clj_diff/miller.clj (hunk 4 above)
Comment (danieldroit) [reply src/clj/clj_diff/miller.clj:82]: cheers

**Review by edwardwas — APPROVED** [review]

**Review by edwardwas — APPROVED** [review]

**Review by dr-egidijus — APPROVED** [review]
LGTM

File: src/clj/clj_diff/core.clj [hunk 5]
```diff
@@ -17,27 +18,32 @@
 
   An index of -1 may appear in additions and is a special case which means to
   add the elements at the beginning of the sequence."
-  [a b]
-  (miller/diff a b))
+  [a b & [opts]]
+  (miller/diff a b opts))
```
Comment (sgepigon) [inline src/clj/clj_diff/core.clj:22]: I'd rather not use varargs.

```suggestion
  ([a b]
   (diff a b {}))
  ([a b opts]
   (miller/diff a b opts)))
```

See https://stuartsierra.com/2015/06/01/clojure-donts-optional-arguments-with-varargs for rationale.

NOTE: This comment applies everywhere else the `& [opts]` idiom is used in this PR.

File: src/clj/clj_diff/core.clj [hunk 6]
```diff
@@ -1,9 +1,10 @@
 (ns clj-diff.core
   "Diff, patch and related functions for Clojure sequences."
-  (:require [clj-diff [miller :as miller]]))
+  (:require [clj-diff [miller :as miller]]
+            [clojure.core.rrb-vector :as fv]))
```
Comment (sgepigon) [inline src/clj/clj_diff/core.clj:4]: Why `fv` as the alias? Is it because `rrb-vector`'s README uses `fv`? https://github.com/clojure/core.rrb-vector#usage

File: project.clj [hunk 7]
```diff
@@ -1,12 +1,13 @@
...
             :distribution :repo
             :comments "same as Clojure"}
 
+  :dependencies [[bsless/clj-fast "0.0.11"]
+                 [org.clojure/core.rrb-vector "0.1.2"]]
   :source-paths ["src/clj"]
-  :java-source-paths ["src/jvm"]
   :plugins [[lein-marginalia "0.9.1"]]
-  :profiles {:dev {:dependencies [[org.clojure/clojure "1.7.0"]]}})
+  :profiles {:dev {:dependencies [[org.clojure/clojure "1.8.0"]]}})
```
Comment (sgepigon) [inline project.clj:13]: Why the bump in the `:dev` profile? And why only to 1.8.0?

Locally, bumping it back down to 1.7.0, I get:
```clj
Exception in thread "main" java.lang.RuntimeException: No such var: clojure.string/starts-with?, compiling:(clj_fast/util.clj:134:14)
```


So IIUC, it's `clj-fast` that requires at least 1.8.0? If that's the case we should declare clojure 1.8.0 explicitly in `:dependencies`.

File: project.clj [hunk 8]
```diff
@@ -1,12 +1,13 @@
-(defproject tech.droit/clj-diff "1.0.1"
+(defproject tech.droit/clj-diff "1.1.0"
```
Comment (sgepigon) [inline project.clj:1]: Please don't bump directly to a release in a feature PR.
```suggestion
(defproject tech.droit/clj-diff "1.1.0-SNAPSHOT"
```
I'd prefer snapshotting to avoid pushing a broken release by mistake.

File: src/clj/clj_diff/miller.clj [hunk 9]
```diff
@@ -41,31 +49,41 @@
...
-                 (get delta))
-             n)
-        [(dec p) delta fp]
-        (recur (inc p)
-               (assoc fp p
-                      (search-p-band a b n m delta p (get fp (dec p) {}))))))))
+
+      (cond
+        (and limit-p
+             (> p limit-p))
```
Comment (sgepigon) [inline src/clj/clj_diff/miller.clj:76]: While this definitely catches the case where `limit-p` is `nil`, perhaps this should also guard if `limit-p` is a number?
```suggestion
        (and (number? limit-p)
             (> p limit-p))
```

Otherwise something like this could be hard to debug:
```clj
user=> (> 10 "1")
Execution error (ClassCastException) at user/eval3 (REPL:1).
class java.lang.String cannot be cast to class java.lang.Number (java.lang.String and java.lang.Number are in module java.base of loader 'bootstrap')
```

File: src/clj/clj_diff/miller.clj [hunk 10]
```diff
@@ -126,21 +144,23 @@
...
+        (recur (dec x) (dec y))))))
 
 ;; See the paper for an example of how there are multiple shortest
 ;; paths through an edit graph.
 
 (defn- next-edit
   "Find the next move through the edit graph which will decrease the
   edit distance by 1."
-  [a b graph delta p x k]
+  [opts a b graph delta p x k]
```
Comment (sgepigon) [inline src/clj/clj_diff/miller.clj:163]: Why is this at beginning of the argslist instead of the end like your other calls?

Also `opts` is unused — is that intended?

File: src/clj/clj_diff/miller.clj [hunk 11]
```diff
@@ -198,18 +218,18 @@
   (map #(vec (cons ::sentinel %)) more))
 
 (defn order->ses
-  [a b]
+  [a b opts]
```
Comment (sgepigon) [inline src/clj/clj_diff/miller.clj:221]: Why are these `opts` not optional? I'm uncomfortable changing the arity of public functions as that's a breaking change.

This comment also applies to `seq-diff`, `diff`, `seq-edit-dist`, `edit-distance`, `seq-lcs`, `longest-common-subseq`.

File: test/clj_diff/test/core.clj [hunk 12]
```diff
@@ -32,7 +32,7 @@
 
 (deftest patch-test
   (is (= (patch "aba"
-                {:+ [[1 \c]], :- [1]})
+                {:+ [[0 \c]], :- [1]})
```
Comment (sgepigon) [inline test/clj_diff/test/core.clj:35]: For my own understanding, why change the test? Both `[0 \c]` `[1 \c]` do `patch` to `"aca"`. 

Removing the `:-` and only patching with the `:+` we get:
```clj
  (patch "aba" {:+ [[0 \c]]}) ;; "acba"
  (patch "aba" {:+ [[1 \c]]}) ;; "abca"
```

At least conceptually `[[1 \c]]` gets the intention across better as you want to replace the character at index 1. If I want to replace the character at index 0, then I'd do `{:+ [[0 \c]] :-[0]}`.

**Review by sgepigon — COMMENTED** [review]
There are some issues I'd like to be addressed before this is released and deployed.

File: src/jvm/clj_diff/FastStringOps.java [hunk 13]
```diff
@@ -1,37 +0,0 @@
-package clj_diff;
-
-/**
- * Fast string operations for clj-diff.
- */
-public class FastStringOps {
```
Comment (MrMikeRodriguez) [inline src/jvm/clj_diff/FastStringOps.java:6]: Removing a class would also be a breaking change.

File: project.clj (hunk 8 above)
Comment (danieldroit) [reply project.clj:1]: I've set it as `2.0.0-experimental` as per @MrMikeRodriguez https://droitfintech.slack.com/archives/C0HEZGGJG/p1660658647737729

File: src/clj/clj_diff/miller.clj (hunk 10 above)
Comment (danieldroit) [reply src/clj/clj_diff/miller.clj:163]: the reason for having it at the beginning was the use with `partial`, but due course the usage of the opts has been eliminated. I'm removing it now from the fn signature.

File: test/clj_diff/test/core.clj (hunk 12 above)
Comment (danieldroit) [reply test/clj_diff/test/core.clj:35]: yes, indeed.

File: src/jvm/clj_diff/FastStringOps.java (hunk 13 above)
Comment (danieldroit) [reply src/jvm/clj_diff/FastStringOps.java:6]: yes, it is thus the new version indicates API change `2.0.0-experimental`

File: src/clj/clj_diff/miller.clj (hunk 9 above)
Comment (danieldroit) [reply src/clj/clj_diff/miller.clj:76]: ok

File: src/clj/clj_diff/miller.clj (hunk 11 above)
Comment (danieldroit) [reply src/clj/clj_diff/miller.clj:221]: done

File: test/clj_diff/test/core.clj (hunk 12 above)
Comment (danieldroit) [reply test/clj_diff/test/core.clj:35]: done

File: src/clj/clj_diff/core.clj (hunk 5 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj:22]: done

File: project.clj (hunk 7 above)
Comment (danieldroit) [reply project.clj:13]: done

File: src/clj/clj_diff/core.clj (hunk 5 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj:22]: yes. fv == FAST VECTOR

File: src/clj/clj_diff/core.clj (hunk 6 above)
Comment (danieldroit) [reply src/clj/clj_diff/core.clj:4]: yes. fv == FAST VECTOR

File: project.clj (hunk 7 above)
Comment (danieldroit) [reply project.clj:13]: done
